### PR TITLE
Adding Pressure BC for multi-app runs

### LINKED ISSUE #11558: Add a pressure boundary condition for codes coupled with multi-apps
## Rationale

When coupling tensor mechanics with flow codes in multi-app setup, we have a need to bring the pressure variable from an external code - typically a multi app.

## Description

Add a boundary condition class that applies pressure in a given direction where the pressure comes from an auxiliary variable.

## Impact

Gain functionality

### PR DESCRIPTION
**Pull request by andrsd**
In coupled runs, the external app computes pressure which is then
transferred into an auxiliary variable. In order to use this externally
computed variable this BC use the variable to impose pressure BC.

Closes #11558

### CODE REVIEW

**Comment by andrsd** [comment]
Tagging @bwspenc and @sapitts

File: modules/tensor_mechanics/test/tests/pressure/error.i [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+[AuxVariables]
+  [./pressure]
+  [../]
+[]
+
+[Kernels]
+  [./TensorMechanics]
+    displacements = 'disp_x disp_y disp_z'
+  [../]
+[]
```
Comment (sapitts) [inline modules/tensor_mechanics/test/tests/pressure/error.i]: You can use the tensor mechanics master action instead of the kernel and strain here, which ensures that we have consistency in the stress divergence use_displaced_mesh setting and the strain formulation

File: modules/tensor_mechanics/test/tests/pressure/pressure_aux.i [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    boundary = 3
+    function = rampUnramp
+  [../]
+[]
+
+[Kernels]
+  [./TensorMechanics]
+    displacements = 'disp_x disp_y disp_z'
+  [../]
+[]
```
Comment (sapitts) [inline modules/tensor_mechanics/test/tests/pressure/pressure_aux.i]: Use the master action here too, and remove the variables, kernel, and strain block

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/Pressure/PressureAction.md [hunk 3]
```diff
@@ -14,18 +14,25 @@ The Pressure Action is used to construct only the [Pressure boundary condition](
...
+- For the variable `disp_z`, the parameter setting `component = 2` is used
 
 Note that the placement of the variables in the `displacements` string determines the value of the component.
 
-As in the [Pressure](/BCs/Pressure.md) boundary condition, the  magnitude of the `Pressure` boundary condition can be specified as either a scalar (use the input parameter `factor`) or a `function` parameter.
+The [Pressure](/BCs/Pressure.md) boundary condition can be prescribed by
+
+1. the magnitude of the `Pressure` using a scalar value (use the input parameter `factor`),
+1. a function (use `function` parameter), or
+1. auxiliary variable (use `pressure` parameter and supply the variable name).
```
Comment (sapitts) [inline modules/tensor_mechanics/doc/content/documentation/systems/BCs/Pressure/PressureAction.md]: Why always `1.` here?

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+The magnitude of the `Pressure` boundary condition has to be computed as an variable and coupled into this class.
+
+A set of `Pressure` boundary conditions applied to multiple variables in multiple components can be defined with the [PressureAction](/BCs/Pressure/PressureAction.md).
+
+!syntax list /BCs/Pressure objects=True actions=False subsystems=False
+
+!syntax list /BCs/Pressure objects=False actions=True subsystems=False
+
+!syntax list /BCs/Pressure objects=False actions=False subsystems=True
```
Comment (sapitts) [inline modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: Please use the non action include syntax for your new class here
```
!syntax parameters /BCs/CoupledPressureBC

!syntax inputs /BCs/CoupledPressureBC

!syntax children /BCs/CoupledPressureBC
```

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+## Description
+
+The boundary condition `CoupledPressureBC` applies a force computed in a variable to a mesh boundary.
+A `component` of the normal vector to the mesh surface (0, 1, or 2 corresponding to the $\hat{x}$, $\hat{y}$, and $\hat{z}$ vector components) is used to determine the direction in which to apply the traction.
+The boundary condition is always applied to the displaced mesh.
+
+The magnitude of the `Pressure` boundary condition has to be computed as an variable and coupled into this class.
+
+A set of `Pressure` boundary conditions applied to multiple variables in multiple components can be defined with the [PressureAction](/BCs/Pressure/PressureAction.md).
+
```
Comment (sapitts) [inline modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: Add the relevant section from an example input file here (with a second level heading)

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
+# Coupled Pressure BC
+
+## Description
+
+The boundary condition `CoupledPressureBC` applies a force computed in a variable to a mesh boundary.
+A `component` of the normal vector to the mesh surface (0, 1, or 2 corresponding to the $\hat{x}$, $\hat{y}$, and $\hat{z}$ vector components) is used to determine the direction in which to apply the traction.
+The boundary condition is always applied to the displaced mesh.
+
+The magnitude of the `Pressure` boundary condition has to be computed as an variable and coupled into this class.
```
Comment (sapitts) [inline modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: Add an equation to demonstrate the calculation that your class computes here please. I agree that it's straightforward, but sometimes an (equation) picture is worth a thousand words, especially to newer users trying to find their footing

**Review by sapitts — COMMENTED** [review]
Just a few requests to make the documentation more uniform with other changes we're trying to put into tensor mechanics

File: modules/tensor_mechanics/test/tests/pressure/error.i (hunk 1 above)
Comment (andrsd) [reply modules/tensor_mechanics/test/tests/pressure/error.i]: I thought this was the action. Can you point me to an example what do you want me to use? I took this from `pressure_test.i`.

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/Pressure/PressureAction.md (hunk 3 above)
Comment (andrsd) [reply modules/tensor_mechanics/doc/content/documentation/systems/BCs/Pressure/PressureAction.md]: MOOSEDocs will render this as 1., 2. and 3 (check for yourself when the doco is generated). The advantage of this is that if you rearrange or add, you do not have to renumber things...

File: modules/tensor_mechanics/test/tests/pressure/error.i (hunk 1 above)
Comment (sapitts) [reply modules/tensor_mechanics/test/tests/pressure/error.i]: This [test](https://github.com/idaholab/moose/blob/devel/modules/tensor_mechanics/test/tests/thermal_expansion/constant_expansion_coeff.i) is pretty close to what you want to use, and we have a small bit of documentation on the master action [here](http://www.mooseframework.org/moose/documentation/systems/Modules/TensorMechanics/Master/index.html) (more improvements coming soon!)

Turns out there is a much wider use of the legacy tensor mechanics action, that is restricted to just kernels, in our test repository than I'd like. Another issue for another day to make all of the tests use the master action.

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/Pressure/PressureAction.md (hunk 3 above)
Comment (sapitts) [reply modules/tensor_mechanics/doc/content/documentation/systems/BCs/Pressure/PressureAction.md]: Well that's pretty neat!

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md (hunk 6 above)
Comment (andrsd) [reply modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: This class will be used in 95% of cases in a multi-app scenario. The pressure will be computed for example by a flow code, like RELAP-7. There we have 2 pressures, we compute a mixture pressure and that is transferred into aux-variable. Then this aux is used by this BC.  Other case is the SAM code which computes pressure in a totally different way (they actually solve for pressure). But the use case is the same. It is transferred into an aux and used by this BC.  That's why I did not put more specific details. But I could describe this here...

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md (hunk 5 above)
Comment (andrsd) [reply modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: ok

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md (hunk 4 above)
Comment (andrsd) [reply modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: Thanks for catching this - that's a copy'n'paste error.

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# Coupled Pressure BC
+
+## Description
+
+The boundary condition `CoupledPressureBC` applies a force computed in a variable to a mesh boundary.
+A `component` of the normal vector to the mesh surface (0, 1, or 2 corresponding to the $\hat{x}$, $\hat{y}$, and $\hat{z}$ vector components) is used to determine the direction in which to apply the traction.
+The boundary condition is always applied to the displaced mesh.
```
Comment (sapitts) [inline modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: Put this in an alert please

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md (hunk 6 above)
Comment (sapitts) [reply modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: Ah, well, in that case, let's leave it as it is and add a short note about the intended use being for multi-apps. The details of the actual use case would probably too confusing for the brand new user

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md (hunk 7 above)
Comment (andrsd) [reply modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: You mean to use  `!alert note`?

Comment (sapitts) [reply modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: yes please!

Comment (andrsd) [reply modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: done

**Comment by andrsd** [comment]
sabertooth failed, but patch for it is already ready...

File: modules/tensor_mechanics/src/actions/PressureAction.C [hunk 8]
```diff
@@ -42,11 +42,18 @@ validParams<PressureAction>()
...
   params.addParam<Real>("alpha", 0.0, "alpha parameter for HHT time integration");
   params.addParam<FunctionName>("function", "The function that describes the pressure");
+  params.addParam<AuxVariableName>("pressure", "The auxiliary variable that contains the pressure");
   return params;
 }
 
 PressureAction::PressureAction(const InputParameters & params) : Action(params)
 {
+  if (isParamValid("pressure") && isParamValid("function"))
+    mooseError("BCs/Pressure/",
```
Comment (permcody) [inline modules/tensor_mechanics/src/actions/PressureAction.C]: Rather than hardcode this, perhaps use paramError() instead? You'd only be able to point to one of the two troublesome params but that would probably be cleaner and give more context to the error message (e.g. the line number).

File: modules/tensor_mechanics/test/tests/pressure/tests [hunk 9]
```diff
@@ -11,4 +11,24 @@
...
     prereq = '3D'
   [../]
+
+  [./pressure_aux]
+    type = Exodiff
+    input = 'pressure_aux.i'
+    exodiff = 'pressure_test_out.e'
+    custom_cmp = 'pressure_aux.exodiff'
+    prereq = '3D_Bbar'
+    requirement = "The system shall allow to apply a pressure boundary condition from an auxiliary variable"
```
Comment (permcody) [inline modules/tensor_mechanics/test/tests/pressure/tests]: 👍

File: modules/tensor_mechanics/src/bcs/CoupledPressureBC.C [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+      "Applies a pressure from an auxiliary variable on a given boundary in a given direction");
+  params.addRequiredParam<unsigned int>("component", "The component for the pressure");
+  params.addRequiredCoupledVar("pressure", "Coupled value containing the pressure");
+  params.set<bool>("use_displaced_mesh") = true;
+  return params;
+}
+
+CoupledPressureBC::CoupledPressureBC(const InputParameters & parameters)
+  : IntegratedBC(parameters),
+    _component(getParam<unsigned int>("component")),
```
Comment (permcody) [inline modules/tensor_mechanics/src/bcs/CoupledPressureBC.C]: Can we do rangeChecked() or some other input validation?

File: modules/tensor_mechanics/include/bcs/CoupledPressureBC.h [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Pressure boundary condition using coupled variable to apply pressure in a given direction
+ */
+class CoupledPressureBC : public IntegratedBC
+{
+public:
+  CoupledPressureBC(const InputParameters & parameters);
+
+protected:
+  virtual Real computeQpResidual();
```
Comment (permcody) [inline modules/tensor_mechanics/include/bcs/CoupledPressureBC.h]: override

**Review by permcody — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/src/actions/PressureAction.C (hunk 8 above)
Comment (andrsd) [reply modules/tensor_mechanics/src/actions/PressureAction.C]: > You'd only be able to point to one of the two troublesome params

Yeah, that's why went with `mooseError`. I can do `paramError()` instead...

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+The boundary condition `CoupledPressureBC` applies a force computed in a variable to a mesh boundary.
+A `component` of the normal vector to the mesh surface (0, 1, or 2 corresponding to the $\hat{x}$, $\hat{y}$, and $\hat{z}$ vector components) is used to determine the direction in which to apply the traction.
+
+!alert note
+The boundary condition is always applied to the displaced mesh.
+
+The `CoupledPressureBC` is typically used in a multi-app scenario.
+The pressure variable can be computed by a sub-app (it can be for example a flow code) and then transferred into an auxiliary variable, which is then coupled into this boundary condition so that it is applied in the master app.
+
+A set of `Pressure` boundary conditions applied to multiple variables in multiple components can be defined with the [PressureAction](/BCs/Pressure/PressureAction.md).
```
Comment (bwspenc) [inline modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: Did you mean 'A set of \`CoupledPressureBC\` boundary conditions`?

File: modules/tensor_mechanics/src/actions/PressureAction.C [hunk 13]
```diff
@@ -42,11 +42,17 @@ validParams<PressureAction>()
   params.addParam<Real>("factor", 1.0, "The factor to use in computing the pressure");
   params.addParam<Real>("alpha", 0.0, "alpha parameter for HHT time integration");
   params.addParam<FunctionName>("function", "The function that describes the pressure");
+  params.addParam<AuxVariableName>("pressure", "The auxiliary variable that contains the pressure");
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/actions/PressureAction.C]: I'm not crazy about this `pressure` parameter name because it doesn't jump out at me as being the option that you would use to specify pressure via a coupled variable.

File: modules/tensor_mechanics/src/bcs/CoupledPressureBC.C [hunk 14]
```diff
@@ -0,0 +1,39 @@
+//* This file is part of the MOOSE framework
```
Comment (bwspenc) [inline modules/tensor_mechanics/src/bcs/CoupledPressureBC.C]: Does this need to be a separate class from the regular Pressure class? We're already doing a couple of switches between different ways of applying it in there anyway?

**Review by bwspenc — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/src/bcs/CoupledPressureBC.C (hunk 14 above)
Comment (andrsd) [reply modules/tensor_mechanics/src/bcs/CoupledPressureBC.C]: `Pressure` has that `factor`, `function` and `postprocessor` stuff which IMO does not make sense in this context. Again, the pressure variable will be computed by an external code in 95% of cases and transferred into.

File: modules/tensor_mechanics/src/actions/PressureAction.C (hunk 13 above)
Comment (andrsd) [reply modules/tensor_mechanics/src/actions/PressureAction.C]: Then, pick a better name. I do not really care. I was going for the syntax that would look like variable coupling...

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md (hunk 12 above)
Comment (andrsd) [reply modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: To me, `Pressure` here indicated type of the BC, rather then C++ class name, but if you think it should read `CoupledPressureBC`, I can change that.

File: modules/tensor_mechanics/src/bcs/CoupledPressureBC.C (hunk 14 above)
Comment (andrsd) [reply modules/tensor_mechanics/src/bcs/CoupledPressureBC.C]: Also, I personally prefer classes that do just one thing rather then a lot of things

Comment (bwspenc) [reply modules/tensor_mechanics/src/bcs/CoupledPressureBC.C]: Sure, I understand that. However, the Pressure class already does multiple things, and this would be a minor addition. 

We probably don't actually need `factor` and `function` because you can prescribe a constant value for `function`. The same is true for the coupled variable. I'd just like to discuss this for a minute.

File: modules/tensor_mechanics/src/actions/PressureAction.C (hunk 13 above)
Comment (bwspenc) [reply modules/tensor_mechanics/src/actions/PressureAction.C]: How about `coupled_pressure`? Also, the description should be changed to drop 'auxiliary' because it could be a solution variable.

File: modules/tensor_mechanics/src/bcs/CoupledPressureBC.C (hunk 14 above)
Comment (bwspenc) [reply modules/tensor_mechanics/src/bcs/CoupledPressureBC.C]: Turns out I'm not going to be around today. I see your point, but I don't think that pattern is followed all that much actually, and we would have a massive number of classes if we actually did follow it everywhere. I'm fine with how you have it, but I'd prefer we go all one way or the other. Following this line of thinking, it seems like we should have a separate Action and syntax block for CoupledPressure as well...

File: modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md (hunk 12 above)
Comment (bwspenc) [reply modules/tensor_mechanics/doc/content/documentation/systems/BCs/CoupledPressureBC.md]: I would prefer it be CoupledPressureBC.

File: modules/tensor_mechanics/src/bcs/CoupledPressureBC.C (hunk 14 above)
Comment (andrsd) [reply modules/tensor_mechanics/src/bcs/CoupledPressureBC.C]: I am fine with doing a different syntax. That's cleaner than having it all in one action.

Something like?
```
[BCs]
  [./CoupledPressure]
    displacements = '...'
    boundary = '<bnd>'
    coupled_pressure = <variable name>
  [../]
[]
```

Here, it looks like `coupled_` prefix is redundant, because it is in the block name. I do not want to go with `variable`, because that could be misinterpreted as the variable this is applied to. What do you think?

Comment (bwspenc) [reply modules/tensor_mechanics/src/bcs/CoupledPressureBC.C]: Sure, if you bust this out into its own syntax block, then I think `pressure` is better than `coupled_pressure`.

Comment (bwspenc) [reply modules/tensor_mechanics/src/bcs/CoupledPressureBC.C]: Same goes for the BC itself.

**Review by bwspenc — APPROVED** [review]
This looks fine to me. I think I would really prefer a design where you just add an option to the existing Pressure BC and Action to handle this extra case. This is just because this adds an awful lot of code and documentation, a lot of which is duplicated from the regular Pressure BC. However, if this is the preferred way to do this kind of thing, go for it.

**Comment by bwspenc** [comment]
@permcody Since you've requested changes, could you please look at this again to see if you are happy with it?

**Review by permcody — APPROVED** [review]

**Comment by permcody** [comment]
This looks good to me but will require a sabertooth patch. @andrsd feel free to merge and make the appropriate changes to sabertooth.
